### ISSUE TITLE
Input lag and performance issues

### ISSUE THREAD

**Issue by goofyseeker311** [hmlendea/geforcenow-electron#31]
quake II rtx and scrap mechanic, while fully working in windows, in the linux version has major unplayable input lag, maybe a minor video lag, tested on windows 8.1, same hardware

**Comment by maciejmrozinski** [comment]
I've observed the same issue on many other games when using Geforce Now inside Electron (previously with Lutris web runner) and when tried to play in Chrome with changed UserAgent to ChromeOS (to be able to launch geforce now). I have constantly growing input and video lag after launching game menus or inside missions. This is probably some issue with Linux libs, but I don't have any clue what could be wrong.

**Comment by hmlendea** [comment]
Firstly, could you please provide the distro, GPU model and drivers you're using?

Secondly, does GFN work well when using it directly through any browser, and if so, how well? (Preferably without using a custom CrOS UserAgent)

**Comment by goofyseeker311** [comment]
ubuntu 20.04.1 LTS, 610M, tried both X and meta-nvidia

i have not tried with a web browser, just the snap

**Comment by hmlendea** [comment]
I also have a 610M, but I'm on Arch Linux.

Would you try the non-snap version? (e.g. the zip version, or deb since you're on Ubuntu)
Let's see if this one works better

**Comment by goofyseeker311** [comment]
sorry, i don't have it installed anymore

pe 8. tammik. 2021 klo 16.23 Horațiu Mlendea (notifications@github.com)
kirjoitti:

**Comment by tosziro** [comment]
I have same problem on Ubuntu 20.04, while using both, browser and electron app (deb version). On same machine works well with original GeForce NOW app on Windows.

**Comment by psywolf85** [comment]
Using Wayland (Mate in Mint) fixed it for me.

**Comment by tosziro** [comment]
> Using Wayland (Mate in Mint) fixed it for me.

Thank you for this tip. I changed Ubuntu session to Wayland and yeah, it works better now.

**Comment by MirceaMx** [comment]
Manjaro - KDE. Intel GPU. Laptop. The SNAP version by another developer does not seem to do this even though it's full-screen implementation is that with borders all-around so I suspect its the full-screen implementation here.
Sure hope the developer finds the solution. Till then: back to Windows.

**Comment by hmlendea** [comment]
Could you guys also point the specific games that you have tried? And also, on a high level, how bad the lag is for each one of you?

For example for me in Black Mesa the lag is slight, but since it's a first-person shooter, it's unenjoyable (but "playable"). In Crusader Kings 3 it's all good since it's a strategy game.

I'm interested to know if for any of you the lag is so huge that nothing's usable.

@MirceaMx  If you run GFN directly through a web browser (specifically Chrome/Chromium) how does it perform?
Also, how does it perform on Windows through the same browser?

**Comment by MirceaMx** [comment]
I am playing The Witcher 3 - GOTY. 
In Chrome on Windows and Linux runs great no problems. Actually thanks on Linux the extension for Chrome runs better than your app and it does not have borders.

**Comment by seitbekir** [comment]
> Uninstalled linux from dual-boot. I was sorry to do that but this os has no chance.
> linux
> OS

Let me ask, have you payed for Windows you are using? Did you install video drivers? Do you know meaning of "OS"?

Related to lag, I had same problems before using macOS with "native" application. So it is issue on GFN side. To fix that you may increase app priority if it is set to low. This helps a lot.

**Comment by MirceaMx** [comment]
_"Let me ask, have you payed for Windows you are using?"_ 
My Laptop has the key integrated in BIOS so in a way yes.

_Did you install video drivers?_
I tried that but since I'm using Manjaro and have a newer Intel GPU I foun out they are already installed.

_Do you know meaning of "OS"?_
OS means Operating System.

**Comment by goofyseeker311** [comment]
I use what I use but I got mine as gift

----- Alkuperäinen viesti -----
Lähettäjä: "MirceaMx" <notifications@github.com>
Lähetetty: ‎2.‎2.‎2021 19:05
Vastaanottaja: "hmlendea/geforcenow-electron" <geforcenow-electron@noreply.github.com>
Kopio: "goofyseeker311" <goofyseeker311@gmail.com>; "Author" <author@noreply.github.com>
Aihe: Re: [hmlendea/geforcenow-electron] Input lag is yuge (#31)

"Let me ask, have you payed for Windows you are using?"
My Laptop has the key integrated in BIOS so in a way yes.
Did you install video drivers?
I tried that but since I'm using Manjaro and have a newer Intel GPU I foun out they are already installed.
Do you know meaning of "OS"?
OS means Operating System.

**Comment by seitbekir** [comment]
So, after I've played for 2h. 1 hour on Mac (native app) and 1 hour on Linux I have a conclusion. Linux performs better. Seems, it is because I have Ryzen 3400G (playing on the integrated Graphics). Also, it played fine via gamepad DS4 on both machines. As using 4K OLED TV I found it is very poor quality for playing, but the traffic was not as high I've expected (I've allowed maximum). The only input lag issue I found it was because of poor connection messages (once in ~5 mins). I believe it is an issue on GFN side only.

Using Manjaro with Gnome and Wayland.

**Comment by ilovespicynoodles** [comment]
I'm having the same issue. I've tried Manjaro, PopOS, Debian and Ubuntu. With both free and non-free drivers and integrated graphics card and my Geforce GPU through prime. The issue persists.
Both Wayland and xorg too.
I have also tried through google chrome and chromium, this app and through lutris but it keeps happening.
I really wonder why this happens. Do others that have this issue also run it on a laptop with both integrated intel gpu and a geforce gpu?

**Comment by maciejmrozinski** [comment]
I tried it (World of Tanks, but also experiencing lags in other games) on integrated Intel® HD Graphics 5500 and such variations:
- Windows 10 (GeForce Now app & Chrome browser) - working great, no lags
- Ubuntu 20.04 X11 (Chrome browser & electron app) - input lag horrible
- Ubuntu 20.04 Wayland (Chrome browser & electron app) - input lag not so terrible, but still hard to play, lag comes and goes from time to time

**Comment by hmlendea** [comment]
So far it seems to universally affect Linux users regardless whether they use this app or not, so it is a more general issue that I am unable to fix.

**Comment by MirceaMx** [comment]
Agree. It is a general issue!

**Comment by dubigrasu** [comment]
It may be the lack of video hardware accelerated decoding. Could be something else as well, but video decoding is the most problematic.

While in game press Shift+Ctrl+Alt+F6 to bring up the stats. Pay attention to the third line and note:
```
rtd: round trip delay from client to server
d: decoding time
```
So **rtd** is basically caused by the distance from you to the Nvidia servers, and there's nothing you can do about it. You can change the server from settings, but the one closest to you is already chosen for you, so likely it won't do any good.
I personally have an **rtd** of 28 to 47ms (depending on the time of the day) and around 30ms is playable.

The **d** value represent how much time your systems spends decoding the frames that are incoming, the lower, the better. The decoding can be done in Software (CPU) or hardware (GPU). Which one does the decoding doesn't really matter as long is done fast. However, you do need a fairly powerful CPU to do this in Software mode. On lower CPU systems is preferable to have Hardware decoding enabled and actually working. 
On my hardware accelerated enabled Chromium I have a **d** of 2.5ms, which gives me a fairly enjoyable experience. Above 5ms (with my already rtd of 30ms) ) the lag starts to be very noticeable, and the gameplay is affected.

In this particular case, geforcenow-electron doesn't seem to have Hardware decoding enabled, and it relies only on the CPU to do the job.

TL;DR: Enable the stats as descried above and post the **rtd** and **d** values

**Comment by goofyseeker311** [comment]
If you have windows working and Linux not, it's not that

----- Alkuperäinen viesti -----
Lähettäjä: "dubigrasu" <notifications@github.com>
Lähetetty: ‎21.‎2.‎2021 15:48
Vastaanottaja: "hmlendea/geforcenow-electron" <geforcenow-electron@noreply.github.com>
Kopio: "goofyseeker311" <goofyseeker311@gmail.com>; "Author" <author@noreply.github.com>
Aihe: Re: [hmlendea/geforcenow-electron] Input lag is huge (#31)

It may be the lack of video hardware accelerated decoding. Could be something else as well, but video decoding is the most problematic.
While in game press Shift+Ctrl+Alt+F6 to bring up the stats. Pay attention to the third line and note:
rtd: round trip delay from client to server
d: decoding time
So rtd is basically caused by the distance from you to the Nvidia servers, and there's nothing you can do about it. You can change the server from settings, but the one closest to you is already chosen for you, so likely it won't do any good.
I personally have an rtd of 28 to 47ms (depending on the time of the day) and around 30ms is playable.
The d value represent how much time your systems spends decoding the frames that are incoming, the lower, the better. The decoding can be done in Software (CPU) or hardware (GPU). Which one does the decoding doesn't really matter as long is done fast. However, you do need a fairly powerful CPU to do this in Software mode. On lower CPU systems is preferable to have Hardware decoding enabled and actually working.
On my hardware accelerated enabled Chromium I have a d of 2.5ms, which gives me a fairly enjoyable experience. Above 5ms (with my already rtd of 30ms) ) the lag starts to be very noticeable, and the gameplay is affected.
In this particular case, geforcenow-electron doesn't seem to have Hardware decoding enabled, and it relies only on the CPU to do the job.
TL;DR: Enable the stats as descried above and post the rtd and d values

**Comment by dubigrasu** [comment]
> If you have windows working and Linux not, it's not that
> […](#)
> ----- Alkuperäinen viesti ----- Lähettäjä: "dubigrasu" <notifications@github.com> Lähetetty: ‎21.‎2.‎2021 15:48 Vastaanottaja: "hmlendea/geforcenow-electron" <geforcenow-electron@noreply.github.com> Kopio: "goofyseeker311" <goofyseeker311@gmail.com>; "Author" <author@noreply.github.com> Aihe: Re: [hmlendea/geforcenow-electron] Input lag is huge (#31) It may be the lack of video hardware accelerated decoding. Could be something else as well, but video decoding is the most problematic. While in game press Shift+Ctrl+Alt+F6 to bring up the stats. Pay attention to the third line and note: rtd: round trip delay from client to server d: decoding time So rtd is basically caused by the distance from you to the Nvidia servers, and there's nothing you can do about it. You can change the server from settings, but the one closest to you is already chosen for you, so likely it won't do any good. I personally have an rtd of 28 to 47ms (depending on the time of the day) and around 30ms is playable. The d value represent how much time your systems spends decoding the frames that are incoming, the lower, the better. The decoding can be done in Software (CPU) or hardware (GPU). Which one does the decoding doesn't really matter as long is done fast. However, you do need a fairly powerful CPU to do this in Software mode. On lower CPU systems is preferable to have Hardware decoding enabled and actually working. On my hardware accelerated enabled Chromium I have a d of 2.5ms, which gives me a fairly enjoyable experience. Above 5ms (with my already rtd of 30ms) ) the lag starts to be very noticeable, and the gameplay is affected. In this particular case, geforcenow-electron doesn't seem to have Hardware decoding enabled, and it relies only on the CPU to do the job. TL;DR: Enable the stats as descried above and post the rtd and d values — You are receiving this because you authored the thread. Reply to this email directly, view it on GitHub, or unsubscribe.

On Windows hardware video decoding support is not a problem, unfortunately it still is on Linux.
While in Windows HW decoding can be easily enabled in your browser, on Linux you must go through some hoops.
For regular browsers you can try: https://www.linuxuprising.com/2021/01/how-to-enable-hardware-accelerated.html
But like I said geforcenow-electron does not have hardware decoding support.

**Comment by strexgth** [comment]
I am still a little bit new to the Linux system, and need help on how to use Wayland! Please help!

*I am using Linux Mint 20.1*

**Comment by ilovespicynoodles** [comment]
I've been very busy with it and i think i MIGHT have found a solution.
I'm running Manjaro using xorg. NOT wayland.
What i did is i installed the "intel-media-driver" through the repository and then started chromium using this command:
`chromium --enable-gpu-rasterization --enable-oop-rasterization --enable-zero-copy --ignore-gpu-blocklist --use-gl=desktop --enable-accelerated-video-decode`

When chrome starts, go to 'chrome://gpu' and see if everything is enabled like here: (Vulkan is not important for this.)
![image](https://user-images.githubusercontent.com/36053358/109402811-26817280-7959-11eb-8553-7ffb02b52f3d.png)

If you don't use a intel integrated gpu, check here for what you need to enable hardware acceleration for your gpu. https://wiki.archlinux.org/index.php/Hardware_video_acceleration

I tried btw to use my nvidia gpu but when i use optimus' prime-run to launch chromium using those commands, it kinda bugs out. 
https://wiki.archlinux.org/index.php/NVIDIA_Optimus

The screen will look a little bit dark but it seems to work perfect now if you play through the browser.

I know this solution isn't for the geforcenow-electron app but i still wanted to give the info, hoping maybe there is a way to get this working in the app. I haven't developed anything using electron before so maybe one of the developers knows more?

It seems like hardware acceleration is not automatically enabled for some users and/or they don't have the needed packages installed.

**Comment by maciejmrozinski** [comment]
@ilovespicynoodles 
Tried the same thing on both X11 and Wayland (Ubuntu 20.04) and still have a lag issues. `chrome://gpu` tab shows exactly the same output as yours (I'm using Chrome 88.0.4324.182).

Sad, but I think GeForce Now just doesn't like Linux users... :(

**Comment by dubigrasu** [comment]
> @ilovespicynoodles
> Tried the same thing on both X11 and Wayland (Ubuntu 20.04) and still have a lag issues. `chrome://gpu` tab shows exactly the same output as yours (I'm using Chrome 88.0.4324.182).
> 
> Sad, but I think GeForce Now just doesn't like Linux users... :(

The thing is, just because you have the same output as in the screenshot posted, that doesn't mean you actually have the hardware acceleration working. If your _browser itself_ is not hardware acceleration enabled, no amount of tweaking will make it work, regardless what _chrome://gpu_ shows.
Geforce Now works just fine on Linux if you have all the necessary bits working.

**Comment by ilovespicynoodles** [comment]
> > @ilovespicynoodles
> > Tried the same thing on both X11 and Wayland (Ubuntu 20.04) and still have a lag issues. `chrome://gpu` tab shows exactly the same output as yours (I'm using Chrome 88.0.4324.182).
> > Sad, but I think GeForce Now just doesn't like Linux users... :(
> 
> The thing is, just because you have the same output as in the screenshot posted, that doesn't mean you actually have the hardware acceleration working. If your _browser itself_ is not hardware acceleration enabled, no amount of tweaking will make it work, regardless what _chrome://gpu_ shows.
> Geforce Now works just fine on Linux if you have all the necessary bits working.

You're correct.

@maciejmrozinski i really recommend reading https://wiki.archlinux.org/index.php/Hardware_video_acceleration and making sure you cover every section.
For example in the beginning using the command `$ mpv --hwdec=auto video_filename` showed it was using hw acceleration, But the browser did not work correctly. Make sure you read the wiki page thoroughly.

**Comment by dubigrasu** [comment]
There is a (rather convoluted, but precise) way to check if the stream you're receiving in your browser is hardware decoded or not. Is a bit off topic here, but it might be useful nevertheless:

- Go to:  chrome://webrtc-internals/ and while the game is on, select the latest stream, and then the **RTCInboundRTPVideoStream_***** section.

- At the **decoderImplementation** line it should say **ExternalDecoder**. That means that the stream is decoded in hardware. If is something else (like ffmpeg), then is in software mode.
![Screenshot from 2021-03-03 23-16-20](https://user-images.githubusercontent.com/6303599/109874515-e2071700-7c77-11eb-948d-f5bfb7b463d7.png)

- Other line to check is **[totalDecodeTime/framesDecoded_in_ms]**, That's the time your system spends (latency) decoding the stream, you want this to be as low as possible. This is basically the same thing as the **d:** value I described few posts back.

And if we speak about latency, your browser itself might cause a bit, go to: https://basro.github.io/input-lag-measuring-tool/ and take a look.
Your mouse should closely follow the inner dot/point. You can improve the response time by starting Chrome with the **--disable-frame-rate-limit** flag. This will cause higher CPU usage though, be aware.

**Comment by pcworld** [comment]
It seems it would be possible to enable hardware video decoding in Electron: https://github.com/electron/electron/issues/18942#issuecomment-793855814

**Comment by 07416** [comment]
> It seems it would be possible to enable hardware video decoding in Electron: [electron/electron#18942 (comment)](https://github.com/electron/electron/issues/18942#issuecomment-793855814)

Can the hw decoding only be enabled when compiling an electron app?

@hmlendea Do you see this as viable?

**Comment by hmlendea** [comment]
Looking into it

**Comment by hmlendea** [comment]
Version 1.1.0 released with the hardware acceleration patches and upgraded electron (v12).

On my end I can see a clear improvement in performance, let me know how it behaves in your cases.

**Comment by gloeglm** [comment]
Video lag for me seems to be almost gone now (before I had multi-second hangs regularly). There is however still slight lag (both input and video) in comparison to the windows native client on the same connection. Enough to make playing not really enjoyable. :(

**Comment by pcworld** [comment]
Would it be possible to add a log message that outputs whether hardware acceleration is actually used in playback?
For me the decode times (Ctrl-Alt-Shift-F6, `d` value) seem to be about the same as before. However I'm using the snap package, and it may be possible that hardware decoding doesn't work there yet (with the current packaging), perhaps similar as to how the Chromium snap package does not support hardware decoding yet.

**Comment by hmlendea** [comment]
@pcworld unfortunately I cannot provide assistance with snap as I have next to no idea how it works and someone else is supporting it for me.
However, in order to take advantage of hardware acceleration, your system must have a VAAPI driver installed (`libva` must be present).

e.g. for me, when I use the regular (non-snap) version, I just need to make sure that `libva` is installed (that's how it's named on Arch Linux at least), and it already is as it was required by other packages before.

@gloeglm Yes, unfortunately the native Windows desktop version seems to be the best performing one. This app here uses electron which means it's limited to how well GFN can run on Chrome/Chromium - which is not as well as the Windows desktop one, unforunately.

**Comment by pcworld** [comment]
Regarding snap and AppImage, it seems that electron-builder does not package the required libraries currently, I've opened an issue upstream: https://github.com/electron-userland/electron-builder/issues/5788

**Comment by 07416** [comment]
Has anyone verified if the Flatpak version can run with hw accel? I already deleted my Nvidia account due to general frustration with the Nvidia's Linux support, but if the Flatpak release is acceleration-capable I'd return for sure!

**Comment by 07416** [comment]
^ @hmlendea Could you update the app on Flathub (v1.0.8) as you appear to be the maintainer?

**Comment by hmlendea** [comment]
@07416 I'm not the maintainer there, I'm just stated as the developer. I have no experience at all with flat nor snap and someone else is taking care of those packages.

**Comment by hmlendea** [comment]
I'm going to close this, as the performance issues have been solved.

If the problem resurfaces, feel free to reopen.

**Comment by dubigrasu** [comment]
I can definitely see the hardware acceleration (the Unified Video Decoder line) working with my AMD GPU:




![Screenshot from 2021-04-16 09-46-23](https://user-images.githubusercontent.com/6303599/114984027-f73bac00-9e99-11eb-9c05-66227c905769.png)

Nice job.

**Comment by ExperiBass** [comment]
Coming to report the same issue on my friends all-in-one, lag only starts when geforcenow-electron is open and running a game (tested with space engineers)

https://linux-hardware.org/?probe=56c6d48f6e
